pkg/nemesis/netem_chaos.go

```diff
@@ -0,0 +1,228 @@
+		Action: chaosv1alpha1.DelayAction,
+		Selector: chaosv1alpha1.SelectorSpec{
+			Namespaces: []string{ns},
+			Pods:       map[string][]string{ns: pods},
+		},
+		// default mode
+		Mode: podMode,
+		// default Latency
+		Delay: &chaosv1alpha1.DelaySpec{
+			Latency:     args[0],
+			Correlation: args[1],
+			Jitter:      args[2],
+		},
+	}
+}
+
+func (d delay) defaultTemplate(ns string, pods []string) chaosv1alpha1.NetworkChaosSpec {
+	return d.template(ns, pods, "90ms", "25", "90ms")
+}
+
+type duplicate struct {
+}
+
+func (d duplicate) netemType() chaosv1alpha1.NetworkChaosAction {
+	return chaosv1alpha1.DuplicateAction
+}
+
+func (d duplicate) template(ns string, pods []string, podMode chaosv1alpha1.PodMode, args ...string) chaosv1alpha1.NetworkChaosSpec {
+	if len(args) != 2 {
+		panic("args number error")
+	}
+	return chaosv1alpha1.NetworkChaosSpec{
+		Action: chaosv1alpha1.DuplicateAction,
+		Selector: chaosv1alpha1.SelectorSpec{
+			Namespaces: []string{ns},
+			Pods:       map[string][]string{ns: pods},
+		},
+		// default mode
+		Mode: podMode,
+		// default Latency
+		Duplicate: &chaosv1alpha1.DuplicateSpec{
+			Duplicate:   args[0],
+			Correlation: args[1],
+		},
+	}
+}
+
+func (d duplicate) defaultTemplate(ns string, pods []string) chaosv1alpha1.NetworkChaosSpec {
+	return d.template(ns, pods, "40", "25")
+}
+
+type corrupt struct {
+}
+
+func (c corrupt) netemType() chaosv1alpha1.NetworkChaosAction {
+	return chaosv1alpha1.CorruptAction
+}
+
+func (c corrupt) template(ns string, pods []string, podMode chaosv1alpha1.PodMode, args ...string) chaosv1alpha1.NetworkChaosSpec {
+	if len(args) != 2 {
+		panic("args number error")
+	}
+	return chaosv1alpha1.NetworkChaosSpec{
+		Action: chaosv1alpha1.DuplicateAction,
+		Selector: chaosv1alpha1.SelectorSpec{
+			Namespaces: []string{ns},
+			Pods:       map[string][]string{ns: pods},
+		},
+		// default mode
+		Mode: podMode,
+		// default Latency
+		Corrupt: &chaosv1alpha1.CorruptSpec{
+			Corrupt:     args[0],
+			Correlation: args[1],
+		},
+	}
+}
+
+func (c corrupt) defaultTemplate(ns string, pods []string) chaosv1alpha1.NetworkChaosSpec {
+	return c.template(ns, pods, "40", "25")
+}
+
+type netemChaos interface {
+	netemType() chaosv1alpha1.NetworkChaosAction
+	template(ns string, pods []string, podMode chaosv1alpha1.PodMode, args ...string) chaosv1alpha1.NetworkChaosSpec
+	defaultTemplate(ns string, pods []string) chaosv1alpha1.NetworkChaosSpec
+}
+
+func selectNetem(name string) netemChaos {
+	switch name {
+	case "loss":
+		return loss{}
+	case "delay":
+		return delay{}
+	case "duplicate":
+		return duplicate{}
+	case "corrupt":
+		return corrupt{}
+	default:
+		panic("selectNetem received an unexists tag")
+	}
+}
+
+// Generate will randomly generate a chaos without selecting nodes.
+func (g netemChaosGenerator) Generate(nodes []cluster.Node) []*core.NemesisOperation {
+	nChaos := selectNetem(g.name)
+	ops := make([]*core.NemesisOperation, 1)
```

Comment:
replace 1 with len(nodes), Generate requires returning a len(nodes) operation array

Reply:
I'll done it.